File: management/host.py
from common import *
import network_mounts

ACTIONS = {
    "system.poweroff",
    "system.reboot",
    "service.start",
    "service.stop",
    "service.restart",
    "service.enable",
    "service.disable",
    "updates.refresh",
    "updates.install",
    "nfs.mount",
    "smb.mount",
    "smb.unmount",
    "nfs.unmount",
    "nfs.export",
    "nfs.export-remove",
    "folder.permissions",
}


def service(value):
    require(
        isinstance(value, str) and re.fullmatch(r"[a-zA-Z0-9_.@:-]+\.service", value),
        'Select a systemd service',
    )
    require(
        not value.startswith(("panasms-", "ssh", "systemd-", "dbus", "network", "NetworkManager", "getty")),
        'This service is protected from changes through the panel',
    )
    require(
        command(["systemctl", "show", "--property=LoadState", "--value", value]).strip() == "loaded",
        'Service is not loaded',
    )
    return value


def query(view, target):
    if view == "nfs":
        return export_query()
    if view == "power":
        raw = command(["vcgencmd", "get_throttled"]).strip()
        value = int(raw.split("=")[1], 16)
        return {
            "throttled": raw,
            "undervoltage": bool(value & 1),
            "throttling": bool(value & 4),
            "pastUndervoltage": bool(value & (1 << 16)),
        }
    if view == "services":
        units = json_command(["systemctl", "list-units", "--type=service", "--all", "--output=json"])
        files = json_command(["systemctl", "list-unit-files", "--type=service", "--output=json"])
        states = {row["unit_file"]: row.get("state") for row in files}
        for unit in units:
            unit["enabled"] = states.get(unit["unit"], "unknown")
        return {"services": units}
    if view == "journal":
        args = ["journalctl", "--no-pager", "--output=json", "--lines=300", "--reverse"]
        filters = json.loads(target) if target.startswith("{") else {"unit": target}
        target = filters.get("unit", "")
        priority = filters.get("priority", 7)
        integer(priority, 0, 7)
        args += ["--priority", str(priority)]
        since = filters.get("since", "24h")
        require(since in ("1h", "24h", "7d", "all"), 'Invalid period')
        if since != "all":
            args += ["--since", {"1h": "1 hour ago", "24h": "1 day ago", "7d": "7 days ago"}[since]]
        boot = filters.get("boot", "all")
        require(boot in ("all", "current", "previous"), 'Invalid boot selection')
        if boot != "all":
            args += ["--boot", "0" if boot == "current" else "-1"]
        if target:
            require(bool(re.fullmatch(r"[a-zA-Z0-9_.@:-]+\.service", target)), 'Invalid service')
            args += ["--unit", target]
        rows = []
        for line in command(args).splitlines():
            row = json.loads(line)
            rows.append(
                {
                    "time": row.get("__REALTIME_TIMESTAMP"),
                    "priority": row.get("PRIORITY"),
                    "unit": row.get("_SYSTEMD_UNIT"),
                    "message": str(row.get("MESSAGE", ""))[:8192],
                }
            )
        return {"entries": rows}
    if view == "updates":
        raw = command(["apt-get", "--simulate", "upgrade"], timeout=60)
        rows = [line for line in raw.splitlines() if line.startswith(("Inst ", "Remv "))]
        return {"packages": rows, "rebootRequired": Path("/run/reboot-required").exists()}
    raise Rejected('Unknown system query')


def plan(action, p, user=None):
    if action in network_mounts.ACTIONS:
        return network_mounts.plan(action, p, user)
    if action == "folder.permissions":
        import pwd, grp

        target = export_path(p.get("target"))
        owner = pwd.getpwnam(name(p.get("owner")))
        group = grp.getgrnam(name(p.get("group")))
        require(owner.pw_uid >= 1000, 'Select a regular user')
        require(
            p.get("mode") in ("0700", "0750", "0770", "0755", "0775", "2770", "2775"), 'Invalid permissions'
        )
        fs = json_command(["findmnt", "--json", "--target", str(target), "--output", "FSTYPE"])[
            "filesystems"
        ][0]["fstype"]
        require(fs in ("ext2", "ext3", "ext4", "xfs", "btrfs"), 'This file system does not support the selected Unix permissions')
        s = target.stat()
        state = [s.st_dev, s.st_ino, s.st_uid, s.st_gid, s.st_mode]
        return {
            "target": str(target),
            "details": [
                str(target),
                'Owner: ' + owner.pw_name,
                'Group: ' + group.gr_name,
                'Permissions: ' + p["mode"],
                'Apply only to the selected folder, without recursion',
            ],
            "confirmation": str(target),
            "fingerprint": fingerprint(action, p, state),
        }
    if action in ("nfs.export", "nfs.export-remove"):
        return export_plan(action, p)
    require(action in ACTIONS, 'Unknown system operation')
    target = p.get("target", "")
    state = {}
    details = []
    if action in ("system.poweroff", "system.reboot"):
        target = "NAS"
        state = Path("/proc/sys/kernel/random/boot_id").read_text().strip()
        details = [action]
    elif action.startswith("service."):
        target = service(target)
        state = command(["systemctl", "show", "--property=ActiveState,UnitFileState,LoadState", target])
        details = [target, action.split(".")[1]]
    elif action.startswith("updates."):
        target = 'OS UPDATE'
        state = query("updates", "")
        details = state["packages"] if action == "updates.install" else ['Refresh available package lists']
        require(action != "updates.install" or bool(details), 'No updates available')
    return {
        "target": target,
        "details": details,
        "confirmation": target,
        "fingerprint": fingerprint(action, p, state),
    }


def execute(action, p, user=None):
    if action in ("system.poweroff", "system.reboot"):
        command(["systemd-run", "--unit=panasms-power-request", "--on-active=5s",
                 "--timer-property=AccuracySec=1s", "/usr/bin/systemctl",
                 "--no-block", action.split(".")[1]])
        return {"message": "Power operation scheduled"}
    if action in network_mounts.ACTIONS:
        return network_mounts.execute(action, p, user)
    if action == "folder.permissions":
        import pwd, grp

        target = export_path(p["target"])
        owner = pwd.getpwnam(p["owner"])
        group = grp.getgrnam(p["group"])
        fd = os.open("/", os.O_RDONLY | os.O_DIRECTORY)
        try:
            for part in target.parts[1:]:
                nextfd = os.open(part, os.O_RDONLY | os.O_DIRECTORY | os.O_NOFOLLOW, dir_fd=fd)
                os.close(fd)
                fd = nextfd
            os.fchown(fd, owner.pw_uid, group.gr_gid)
            os.fchmod(fd, int(p["mode"], 8))
        finally:
            os.close(fd)
        return {"message": 'Folder ownership and permissions updated'}
    if action in ("nfs.export", "nfs.export-remove"):
        return export_execute(action, p)
    target = p.get("target", "")
    if action.startswith("service."):
        verb = action.split(".")[1]
        command(["systemctl", verb, target])
        state = command(["systemctl", "show", "--property=ActiveState,UnitFileState", "--value", target])
        return {"message": state}
    if action == "updates.refresh":
        command(["apt-get", "update"], timeout=1800)
    elif action == "updates.install":
        command(["apt-get", "-y", "-o", "Dpkg::Options::=--force-confold", "upgrade"], timeout=86400)
    return {"message": 'Done', "rebootRequired": Path("/run/reboot-required").exists()}


def export_path(value):
    from storage import mountpoint

    point = mountpoint(value)
    require(point.is_dir(), 'Directory does not exist')
    row = json_command(["findmnt", "--json", "--target", str(point), "--output", "TARGET,FSTYPE"])[
        "filesystems"
    ][0]
    require(
        row["target"] != "/" and row["fstype"] not in ("nfs", "nfs4", "cifs", "autofs"),
        'The shared folder must be on a mounted local volume',
    )
    require(
        not any(c in str(point) for c in ' \t\\"'), 'The share path must not contain spaces or quotation marks'
    )
    return point


def export_plan(action, p):
    import ipaddress

    target = export_path(p.get("target"))
    conf = Path("/etc/exports.d/panasms.exports")
    state = conf.read_text() if conf.exists() else ""
    if action == "nfs.export":
        clients = p.get("clients", "").split(",")
        require(0 < len(clients) <= 32, 'Specify NFS clients')
        for client in clients:
            try:
                ipaddress.ip_network(client.strip(), strict=False)
            except ValueError:
                raise Rejected('Specify clients as comma-separated IP addresses or subnets')
    return {
        "target": str(target),
        "details": [
            str(target),
            'Clients: ' + p.get("clients", ""),
            'NFS: Linux UID/GID permissions; root_squash enabled',
        ],
        "confirmation": str(target),
        "fingerprint": fingerprint(action, p, state),
    }


def export_execute(action, p):
    target = str(export_path(p["target"]))
    conf = Path("/etc/exports.d/panasms.exports")
    rows = conf.read_text().splitlines() if conf.exists() else []
    rows = [line for line in rows if not line.startswith(target + " ")]
    if action == "nfs.export":
        rows.append(
            target
            + " "
            + " ".join(
                c.strip()
                + "("
                + ("ro" if p.get("readOnly") else "rw")
                + ",sync,no_subtree_check,root_squash)"
                for c in p["clients"].split(",")
            )
        )
    atomic(conf, "\n".join(rows) + "\n", 0o644)
    command(["systemctl", "enable", "--now", "nfs-kernel-server"])
    command(["exportfs", "-ra"])
    return {"message": 'NFS share updated'}


def export_query():
    path = Path("/etc/exports.d/panasms.exports")
    entries = []
    for line in path.read_text().splitlines() if path.exists() else []:
        fields = line.split()
        if fields and not fields[0].startswith("#"):
            entries.append(
                {
                    "path": fields[0],
                    "clients": [v.split("(")[0] for v in fields[1:]],
                    "readOnly": all("(ro," in v for v in fields[1:]),
                }
            )
    return {"exports": entries}

File: internal/systemops/host/host.go
// Package host is the Go port of backend/management/host.py's system, service,
// journal, package-update and power operations (GO-03). It reproduces the exact
// command arguments, timeouts, protected-service rules and human-readable
// messages of the Python layer so the panel behaves identically.
//
// Deliberately NOT ported here (kept on the legacy Python dispatcher until their
// own tasks own them): folder.permissions and the NFS/SMB export/mount actions
// (network_mounts), which move in GO-06/GO-07. HTTP port changes are owned
// separately by systemops/webaccess.
package host

import (
	"context"
	"encoding/json"
	"regexp"
	"strconv"
	"strings"
	"time"

	"panasms.local/backend/internal/systemops"
)

// serviceUnit matches host.py's service() unit-name rule.
var serviceUnit = regexp.MustCompile(`^[a-zA-Z0-9_.@:-]+\.service$`)

// protectedPrefixes are unit-name prefixes host.py refuses to act on through
// the panel, protecting the OS and PaNasMs itself from panel-driven changes.
var protectedPrefixes = []string{"panasms-", "ssh", "systemd-", "dbus", "network", "NetworkManager", "getty"}

// packageLine matches "Inst "/"Remv " lines emitted by apt-get --simulate.
var packageLine = regexp.MustCompile(`^(Inst|Remv) (\S+)(?: \[([^\]]+)\])?(?: \((\S+) (.*)\))?`)

// service validates and returns a systemd unit name, reproducing host.service():
// it must look like a *.service unit, must not be a protected unit, and must be
// loaded. It runs systemctl to confirm the load state.
func service(ctx context.Context, value string) (string, error) {
	if !serviceUnit.MatchString(value) {
		return "", reject("Select a systemd service")
	}
	for _, p := range protectedPrefixes {
		if strings.HasPrefix(value, p) {
			return "", reject("This service is protected from changes through the panel")
		}
	}
	out, err := systemops.Command(ctx, []string{"systemctl", "show", "--property=LoadState", "--value", value}, systemops.CommandOptions{})
	if err != nil {
		return "", err
	}
	if strings.TrimSpace(string(out)) != "loaded" {
		return "", reject("Service is not loaded")
	}
	return value, nil
}

func reject(msg string) error { return &systemops.Rejected{Message: msg} }

// Query serves the services/journal/power/updates views, mirroring host.query.
func Query(ctx context.Context, view, target string) (json.RawMessage, error) {
	switch view {
	case "nfs":
		// nfs export listing stays legacy (GO-07); this Go path is not routed
		// for it, but guard anyway.
		return nil, reject("Unknown system query")
	case "power":
		return powerQuery(ctx)
	case "services":
		return servicesQuery(ctx)
	case "journal":
		return journalQuery(ctx, target)
	case "updates":
		return updatesQuery(ctx)
	}
	return nil, reject("Unknown system query")
}

func powerQuery(ctx context.Context) (json.RawMessage, error) {
	out, err := systemops.Command(ctx, []string{"vcgencmd", "get_throttled"}, systemops.CommandOptions{})
	if err != nil {
		return nil, err
	}
	raw := strings.TrimSpace(string(out))
	// raw looks like "throttled=0x0"; parse the hex after '='.
	parts := strings.SplitN(raw, "=", 2)
	if len(parts) != 2 {
		return nil, reject("Unknown system query")
	}
	value, err := strconv.ParseUint(strings.TrimPrefix(strings.TrimSpace(parts[1]), "0x"), 16, 64)
	if err != nil {
		return nil, reject("Unknown system query")
	}
	return marshal(map[string]any{
		"throttled":        raw,
		"undervoltage":     value&1 != 0,
		"throttling":       value&4 != 0,
		"pastUndervoltage": value&(1<<16) != 0,
	})
}

func servicesQuery(ctx context.Context) (json.RawMessage, error) {
	unitsRaw, err := systemops.Command(ctx, []string{"systemctl", "list-units", "--type=service", "--all", "--output=json"}, systemops.CommandOptions{})
	if err != nil {
		return nil, err
	}
	filesRaw, err := systemops.Command(ctx, []string{"systemctl", "list-unit-files", "--type=service", "--output=json"}, systemops.CommandOptions{})
	if err != nil {
		return nil, err
	}
	var units []map[string]any
	if err := systemops.Decode(unitsRaw, &units); err != nil {
		return nil, reject("Unknown system query")
	}
	var files []map[string]any
	if err := systemops.Decode(filesRaw, &files); err != nil {
		return nil, reject("Unknown system query")
	}
	states := map[string]any{}
	for _, row := range files {
		if uf, ok := row["unit_file"].(string); ok {
			states[uf] = row["state"]
		}
	}
	for _, unit := range units {
		name, _ := unit["unit"].(string)
		if s, ok := states[name]; ok {
			unit["enabled"] = s
		} else {
			unit["enabled"] = "unknown"
		}
	}
	return marshal(map[string]any{"services": units})
}

func journalQuery(ctx context.Context, target string) (json.RawMessage, error) {
	args := []string{"journalctl", "--no-pager", "--output=json", "--lines=300", "--reverse"}
	filters := map[string]any{}
	if strings.HasPrefix(target, "{") {
		if err := systemops.Decode([]byte(target), &filters); err != nil {
			return nil, reject("Invalid request")
		}
	} else {
		filters["unit"] = target
	}
	unit, _ := filters["unit"].(string)

	priority := int64(7)
	if p, ok := filters["priority"]; ok {
		n, err := toInt(p)
		if err != nil {
			return nil, reject("Number outside the allowed range")
		}
		priority = n
	}
	if _, err := systemops.Integer(priority, 0, 7); err != nil {
		return nil, err
	}
	args = append(args, "--priority", strconv.FormatInt(priority, 10))

	since := "24h"
	if s, ok := filters["since"].(string); ok {
		since = s
	}
	sinceArg := map[string]string{"1h": "1 hour ago", "24h": "1 day ago", "7d": "7 days ago"}
	if since != "all" {
		mapped, ok := sinceArg[since]
		if !ok {
			return nil, reject("Invalid period")
		}
		args = append(args, "--since", mapped)
	}

	boot := "all"
	if b, ok := filters["boot"].(string); ok {
		boot = b
	}
	switch boot {
	case "all":
	case "current":
		args = append(args, "--boot", "0")
	case "previous":
		args = append(args, "--boot", "-1")
	default:
		return nil, reject("Invalid boot selection")
	}

	if unit != "" {
		if !serviceUnit.MatchString(unit) {
			return nil, reject("Invalid service")
		}
		args = append(args, "--unit", unit)
	}

	out, err := systemops.Command(ctx, args, systemops.CommandOptions{})
	if err != nil {
		return nil, err
	}
	rows := []map[string]any{}
	for _, line := range strings.Split(string(out), "\n") {
		if strings.TrimSpace(line) == "" {
			continue
		}
		var row map[string]any
		if json.Unmarshal([]byte(line), &row) != nil {
			continue
		}
		message := ""
		if m, ok := row["MESSAGE"]; ok {
			message = truncate(toString(m), 8192)
		}
		rows = append(rows, map[string]any{
			"time":     row["__REALTIME_TIMESTAMP"],
			"priority": row["PRIORITY"],
			"unit":     row["_SYSTEMD_UNIT"],
			"message":  message,
		})
	}
	return marshal(map[string]any{"entries": rows})
}

func updatesQuery(ctx context.Context) (json.RawMessage, error) {
	packages, details, err := simulateUpgrade(ctx)
	if err != nil {
		return nil, err
	}
	return marshal(map[string]any{
		"packages":       packages,
		"packageDetails": details,
		"rebootRequired": rebootRequired(),
	})
}

// simulateUpgrade runs `apt-get --simulate upgrade` and returns the Inst/Remv
// lines and their parsed detail records, matching host.query("updates").
func simulateUpgrade(ctx context.Context) ([]string, []map[string]any, error) {
	out, err := systemops.Command(ctx, []string{"apt-get", "--simulate", "upgrade"}, systemops.CommandOptions{Timeout: 60 * time.Second})
	if err != nil {
		return nil, nil, err
	}
	packages := []string{}
	details := []map[string]any{}
	for _, line := range strings.Split(string(out), "\n") {
		if strings.HasPrefix(line, "Inst ") || strings.HasPrefix(line, "Remv ") {
			packages = append(packages, line)
			details = append(details, packageDetail(line))
		}
	}
	return packages, details, nil
}

func packageDetail(line string) map[string]any {
	m := packageLine.FindStringSubmatch(line)
	if m == nil {
		return map[string]any{"name": line, "installed": "", "available": "", "source": "", "action": "unknown"}
	}
	action := "install"
	if m[1] == "Remv" {
		action = "remove"
	}
	return map[string]any{"name": m[2], "installed": m[3], "available": m[4], "source": m[5], "action": action}
}

func marshal(v any) (json.RawMessage, error) {
	data, err := json.Marshal(v)
	if err != nil {
		return nil, err
	}
	return json.RawMessage(data), nil
}

func toString(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	data, _ := json.Marshal(v)
	return string(data)
}

func truncate(s string, n int) string {
	runes := []rune(s)
	if len(runes) > n {
		return string(runes[:n])
	}
	return s
}

func toInt(v any) (int64, error) {
	switch n := v.(type) {
	case json.Number:
		return n.Int64()

	}
	return 0, reject("Number outside the allowed range")
}
